keep/api/routes/alerts.py

```diff
@@ -1456,3 +1464,57 @@ def dismiss_error_alerts(
         )
 
         return {"success": True, "message": "Successfully dismissed all alerts"}
+
+
+@staticmethod
+def calculate_fingerprint(alert: dict) -> str:
+    """
+    Calculate fingerprint for alert.
+    Returns the original fingerprint if provided either in labels or at root level,
+    otherwise calculates a SHA256 hash based on labels or alert properties.
+    """
+    # First check labels for fingerprint
+    labels = alert.get("labels", {})
+    fingerprint = labels.get("fingerprint")
+    if fingerprint:
+        logger.debug("Using fingerprint from alert labels")
+        return fingerprint
+
+    # Then check root level fingerprint
+    fingerprint = alert.get("fingerprint")
+    if fingerprint:
+        logger.debug("Using fingerprint from alert root level")
+        return fingerprint
+
+    # If no fingerprint found, calculate one
+    fingerprint_string = None
+    if not labels:
+        logger.warning(
+            "No labels found in alert will use old behaviour",
+            extra={
+                "labels": labels,
+            },
+        )
+    else:
+        try:
+            logger.info(
+                "No fingerprint in alert, calculating fingerprint by labels"
+            )
+            fingerprint_string = json.dumps(labels)
+        except Exception:
+            logger.exception(
+                "Failed to calculate fingerprint",
+                extra={
+                    "labels": labels,
+                },
+            )
+
+    # If no labels or failed to calculate from labels
+    if not fingerprint_string:
+        # old behavior
+        service = GrafanaProvider.get_service(alert)
```

Comment:
You should not call the `GrafanaProvider` here, only operate with provider_class